eessi/testsuite/tests/apps/walberla/walberla.py

```diff
@@ -0,0 +1,152 @@
+The test is taken from the existing tutorials within the walberla repository:
+https://github.com/lssfau/walberla/tree/master/apps/tutorials/lbm
+
+License information can be found here:
+https://github.com/lssfau/walberla/blob/master/COPYING.txt
+
+ReFrame terminology:
+
+"pipeline stages":
+https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html#pipeline-hooks
+
+"test parameter": a list of values, which will generate different test variants.
+https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html#reframe.core.builtins.parameter
+
+"test variant": a version of a test with a specific value for each test parameter
+https://reframe-hpc.readthedocs.io/en/stable/regression_test_api.html#test-variants
+
+"concrete test cases": all test combinations that will actually run:
+- test variants
+- valid system:partition+programming environment combinations
+https://reframe-hpc.readthedocs.io/en/stable/tutorial_deps.html#listing-dependencies
+
+Tests can be filtered by name, tag, programming environment, system, partition, or maintainer,
+see https://reframe-hpc.readthedocs.io/en/stable/manpage.html#test-filtering
+
+Hooks acting on all possible test combinations (before filtering) are called after the 'init' stage.
+Hooks acting on concrete test cases (after filtering) are called after the 'setup' stage.
+
+See also https://reframe-hpc.readthedocs.io/en/stable/pipeline.html
+"""
+
+import reframe as rfm
+# added only to make the linter happy
+from reframe.core.builtins import deferrable, parameter, run_after, run_before, sanity_function, performance_function
+import reframe.utility.sanity as sn
+from eessi.testsuite.constants import COMPUTE_UNITS, DEVICE_TYPES, SCALES
+from eessi.testsuite.eessi_mixin import EESSI_Mixin
+from eessi.testsuite.utils import find_modules
+
+
+def filter_scales():
+    """
+    Filtering function for filtering scales for the waLBerla test returns all scales from 8 cores to 2 nodes. This is
+    because this case is testing strong scaling and the number of cells per MPI process become too less typically once
+    it crosses 2 nodes. Further filtering is done within the test based on MPI tasks. The code is capable of running on
+    1 to 4 cores as well but it will take about 1 hour to run on a single core.
+    """
+    return [
+        k for (k, v) in SCALES.items()
+        if ((v['num_nodes'] <= 2) and (v.get('node_part', 0) != 0)) or (v.get('num_cpus_per_node', 0) > 4)
+    ]
+
+
+@rfm.simple_test
+class EESSI_WALBERLA_BACKWARD_FACING_STEP(rfm.RunOnlyRegressionTest, EESSI_Mixin):
+    """
+    This is the main walberla class for the backward facing step test. The test consists of steps that modify the input
+    file from the precompiled walberla tutorials binary. These are run as pre-run commands and the main test with the
+    executable `05_BackwardFacingStep` is measured for performance.
+    The code will run on one core but will take about an hour to run.
```

Comment:
This should probably be adapted, and just state that this test will only run on scales that correspond to at least 8 cores and at most 384 cores. I think that's nice to say here explicitly because _this_ docstring is the one that people are likely to read if they've tried `reframe -t 4_cores` and are wondering why this test is not instantiated.